pkg/corpus/corpus.go

```diff
@@ -0,0 +1,235 @@
+type ItemUpdate struct {
+	Call     int
+	RawCover []uint32
+}
+
+// Item objects are to be treated as immutable, otherwise it's just
+// too hard to synchonize accesses to them across the whole project.
+// When Corpus updates one of its items, it saves a copy of it.
+type Item struct {
+	Sig      string
+	Call     int
+	Prog     *prog.Prog
+	ProgData []byte // to save some Serialize() calls
+	HasAny   bool   // whether the prog contains squashed arguments
+	Signal   signal.Signal
+	Cover    []uint32
+	Updates  []ItemUpdate
+}
+
+func (item Item) StringCall() string {
+	return stringCall(item.Prog, item.Call)
+}
+
+func (item Item) RPCInput() rpctype.Input {
+	return rpctype.Input{
+		Call:   item.Call,
+		Prog:   item.ProgData,
+		Signal: item.Signal.Serialize(),
+	}
+}
+
+func stringCall(p *prog.Prog, call int) string {
+	if call != -1 {
+		return p.Calls[call].Meta.Name
+	}
+	return ".extra"
+}
+
+type NewInput struct {
+	Prog     *prog.Prog
+	Call     int
+	Signal   signal.Signal
+	Cover    []uint32
+	RawCover []uint32
+}
+
+func (item NewInput) StringCall() string {
+	return stringCall(item.Prog, item.Call)
+}
+
+type NewItemEvent struct {
+	NewInput
+	Sig      string
+	Exists   bool
+	ProgData []byte
+}
+
+func (item NewItemEvent) RPCInput() rpctype.Input {
+	return rpctype.Input{
+		Call:     item.Call,
+		Prog:     item.ProgData,
+		Signal:   item.Signal.Serialize(),
+		Cover:    item.Cover,
+		RawCover: item.RawCover,
+	}
+}
+
+func (corpus *Corpus) Save(inp NewInput) {
+	corpus.mu.Lock()
+	defer corpus.mu.Unlock()
+
+	progData := inp.Prog.Serialize()
+	sig := hash.String(progData)
+
+	update := ItemUpdate{
+		Call:     inp.Call,
+		RawCover: inp.RawCover,
+	}
+	exists := false
+	if old, ok := corpus.progs[sig]; ok {
+		exists = true
+		newSignal := old.Signal.Copy()
+		newSignal.Merge(inp.Signal)
+		var newCover cover.Cover
+		newCover.Merge(old.Cover)
+		newCover.Merge(inp.Cover)
+		newItem := &Item{
+			Sig:      sig,
+			Prog:     old.Prog,
+			ProgData: progData,
+			Call:     old.Call,
+			HasAny:   old.HasAny,
+			Signal:   newSignal,
+			Cover:    newCover.Serialize(),
+			Updates:  append([]ItemUpdate{}, old.Updates...),
+		}
+		const maxUpdates = 32
+		if len(newItem.Updates) < maxUpdates {
+			newItem.Updates = append(newItem.Updates, update)
+		}
+		corpus.progs[sig] = newItem
+	} else {
+		corpus.progs[sig] = &Item{
+			Sig:      sig,
+			Call:     inp.Call,
+			Prog:     inp.Prog,
+			ProgData: progData,
+			HasAny:   inp.Prog.ContainsAny(),
+			Signal:   inp.Signal,
+			Cover:    inp.Cover,
+			Updates:  []ItemUpdate{update},
+		}
+		corpus.saveProgram(inp.Prog, inp.Signal)
+	}
+	corpus.signal.Merge(inp.Signal)
+	if corpus.updates != nil {
+		select {
+		case <-corpus.ctx.Done():
+		case corpus.updates <- NewItemEvent{
+			Sig:      sig,
+			NewInput: inp,
+			Exists:   exists,
+			ProgData: progData,
+		}:
+		}
+	}
+}
+
+func (corpus *Corpus) ReportUpdates(updates chan<- NewItemEvent) {
```

Comment:
There's one tricky case -- `syz-fuzzer` receiving corpus items from `syz-manager`. We don't want to get these updates over the channel and process them, which would happen if we just passed it to the constructor.

I thought we could maybe transform it to

`func (corpus *Corpus) ReportUpdatesC() updates <-chan NewItemEvent`

and call it once we want to start getting these updates. It will save a few lines of code.

Reply:
> We don't want to get these updates over the channel and process them, which would happen if we just passed it to the constructor.

But we start polling in the syz-fuzzer after ReportUpdates call. So we will still receive these updates, no?
We also receive inputs from manager when syz-fuzzer runs IIRC.

Reply:
Hmmm, yes, that's true.

Reply:
I've returned the `NewInputs` channel from the fuzzer, it will only stream new programs (and it will be removed in one of the future PRs). The `Corpus.updates` one is now only used to keep `corpus.db` in sync.